crates/bevy_asset/src/asset_server.rs

```diff
@@ -384,10 +390,36 @@ impl AssetServer {
     }
 
     pub fn free_unused_assets(&self) {
+        let mut potential_frees = self.server.asset_ref_counter.mark_unused_assets.write();
+
+        if !potential_frees.is_empty() {
+            let ref_counts = self.server.asset_ref_counter.ref_counts.read();
+            let asset_sources = self.server.asset_sources.read();
+            let asset_lifecycles = self.server.asset_lifecycles.read();
+            for potential_free in potential_frees.iter() {
+                if let Some(&0) = ref_counts.get(potential_free) {
+                    let type_uuid = match potential_free {
+                        HandleId::Id(type_uuid, _) => Some(*type_uuid),
+                        HandleId::AssetPathId(id) => asset_sources
+                            .get(&id.source_path_id())
+                            .and_then(|source_info| source_info.get_asset_type(id.label_id())),
+                    };
+
+                    if let Some(type_uuid) = type_uuid {
+                        if let Some(asset_lifecycle) = asset_lifecycles.get(&type_uuid) {
+                            asset_lifecycle.free_asset(*potential_free);
+                        }
+                    }
+                }
+            }
+            potential_frees.clear();
+        }
+    }
+
+    pub fn mark_unused_assets(&self) {
         let receiver = &self.server.asset_ref_counter.channel.receiver;
         let mut ref_counts = self.server.asset_ref_counter.ref_counts.write();
-        let asset_sources = self.server.asset_sources.read();
-        let mut potential_frees = Vec::new();
+        let mut potential_frees = self.server.asset_ref_counter.mark_unused_assets.write();
```

Comment:
In the case were we only have `RefChange::Increment` we're unnecessarily holding onto the mutex here. It's considering changing this to be `Option<RwLockWriteGuard>` and then in the case we have a `RefChange::Decrement` actually grabbing exclusive access.

So we would need to change line 435 to something like
```rust
if *entry == 0 {
    potential_frees.get_or_insert_with(|| self.server.asset_ref_counter.makr_unused_assets.write()).push(handle_id);
}
```